### PR TITLE
Use of constrained types instead of asserts.

### PR DESCRIPTION
**Pull request by jaheba**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/gluonts/core/ty.py [hunk 1]
```diff
@@ -100,3 +100,10 @@ def fn_wrapper(*args, **kwargs):
     fn_wrapper.__checked__ = checked_args
 
     return fn_wrapper
+
+
+PositiveInt = pydantic.PositiveInt
+PositiveInt0 = pydantic.conint(ge=0)
```
Comment (lostella) [inline src/gluonts/core/ty.py:106]: `NonnegativeInt` maybe?

Comment (jaheba) [reply src/gluonts/core/ty.py:106]: Let me mark this as a draft :).

**Comment by lostella** [comment]
One advantage of this is that we can get validation in the network constructors too, without duplicating hundreds of lines. This is useful in case one wants to directly handle networks, without going through any estimator class

**Comment by vafl** [comment]
I think we had this at some point (or at least considered doing it). If I remember correctly, we decided against it, because  many IDEs like intellij will show a warning when you pass an integer and don't wrap it in `PositiveInt` explicitly and it didn't really bring that much benefit. We can of course reconsider, but I think the warnings issue would be quite annoying.

**Comment by jaheba** [comment]
> I think we had this at some point (or at least considered doing it). If I remember correctly, we decided against it, because many IDEs like intellij will show a warning when you pass an integer and don't wrap it in `PositiveInt` explicitly and it didn't really bring that much benefit. We can of course reconsider, but I think the warnings issue would be quite annoying.
> 

I really dislike this argument: Some IDE is dictating how to write code, which I feel not comfortable with.

This is a problem of the IDE and not the code.

> Also mypy will probably complain when you pass an `int` to a `PositiveInt` argument.

No, this part works.

**Comment by vafl** [comment]
> I really dislike this argument: Some IDE is dictating how to write code, which I feel not comfortable with.
> This is a problem of the IDE and not the code.

Yeah, I agree its not ideal. I think for pycharm the issue may be this: https://youtrack.jetbrains.com/issue/PY-40569
On the other hand the manual assert is only a bit more verbose. 

>> Also mypy will probably complain when you pass an int to a PositiveInt argument.

> No, this part works.

Ah, cool. Is that because of the `validated` wrapper magic?

**Comment by jaheba** [comment]
> Ah, cool. Is that because of the validated wrapper magic?

There are stub-files (ending in `.pyi`), which take preference over the `.py` file when type-checking:

```
PositiveInt = int
PositiveInt0 = int

PositiveFloat = float
PositiveFloat0 = float
```

That way, mypy assumes that for example `PositiveInt` is just `int`.

**Comment by vafl** [comment]
Ok, nice. Maybe that works with pycharm, too then. I think we tried with some more naive option.

**Comment by jaheba** [comment]
> Ok, nice. Maybe that works with pycharm, too then. I think we tried with some more naive option.

According to the documentation, there is support for stub files in pycharm: https://www.jetbrains.com/help/pycharm/stubs.html

File: src/gluonts/model/deepar/_estimator.py [hunk 2]
```diff
@@ -127,118 +137,63 @@ class DeepAREstimator(GluonEstimator):
...
-            if lags_seq is not None
-            else get_lags_for_frequency(freq_str=freq)
-        )
-        self.time_features = (
-            time_features
-            if time_features is not None
-            else time_features_from_frequency_str(self.freq)
-        )
+    def __init__(self, **fields):
+        BaseModel.__init__(self, **fields)
```
Comment (lostella) [inline src/gluonts/model/deepar/_estimator.py:169]: Can one pass anything here to the constructor? Or just the fields declared in the class?

Comment (jaheba) [reply src/gluonts/model/deepar/_estimator.py:169]: You could pass anything there, but only the fields which are declared are handled.

**Pull request closed without merging**